File: c/22-2.c
#include "common.h"

typedef struct Cube {
  i64 x1, x2, y1, y2, z1, z2;
  bool on;
} Cube;

#define CUBES_CAP 1024
Cube cubes[CUBES_CAP];
int cubes_len = 0;

void parse() {
  i64 x1, y1, z1, x2, y2, z2;
  char mode[5];
  FILE *f = fopen(INFILE("22.txt"), "r");
  while (fscanf(f, " %s x=%lld..%lld,y=%lld..%lld,z=%lld..%lld", 
                mode, &x1, &x2, &y1, &y2, &z1, &z2) != EOF) 
    cubes[cubes_len++] = (Cube){x1, x2, y1, y2, z1, z2, mode[1] == 'n'};
  fclose(f);
}

i64 Cube_area(Cube c) {
  return (c.x2 - c.x1 + 1) * (c.y2 - c.y1 + 1) * (c.z2 - c.z1 + 1);
}

void Cube_print(Cube c) {
  printf("(%lld,%lld,%lld) -> (%lld,%lld,%lld): %d\n", c.x1, c.y1, c.z1, c.x2, c.y2, c.z2, c.on);
}

bool Cube_overlaps(Cube a, Cube b) {
  return  !(a.x2 < b.x1 || a.x1 > b.x2 ||
            a.y2 < b.y1 || a.y1 > b.y2 ||
            a.z2 < b.z1 || a.z1 > b.z2);
}

Cube Cube_overlap(Cube a, Cube b) {
  return (Cube){
    .x1 = max(a.x1, b.x1),
    .y1 = max(a.y1, b.y1),
    .z1 = max(a.z1, b.z1),
    .x2 = min(a.x2, b.x2),
    .y2 = min(a.y2, b.y2),
    .z2 = min(a.z2, b.z2),
    .on = !b.on
  };
}

i64 solve(bool inbounds) {
  static Cube disjoint[CUBES_CAP * CUBES_CAP];
  int disjoint_len = 0;

  disjoint_len = 0;
  for (int i = 0; i < cubes_len; i++) {
    static const Cube bounds = {-50, 50, -50, 50, -50, 50, false};
    if (inbounds && !Cube_overlaps(cubes[i], bounds)) continue;

    int N = disjoint_len;
    for (int j = 0; j < N; j++)
      if (Cube_overlaps(cubes[i], disjoint[j]))
        disjoint[disjoint_len++] = Cube_overlap(cubes[i], disjoint[j]);
    if (cubes[i].on) 
      disjoint[disjoint_len++] = cubes[i];
  }
  i64 sum = 0;
  for (int i = 0; i < disjoint_len; i++)
    sum += Cube_area(disjoint[i]) * (disjoint[i].on ? 1 : -1);
  return sum;
}

int main() {
  parse();
  printf("Part 1: %lld\n", solve(true));
  printf("Part 2: %lld\n", solve(false));
  return 0;
}

File: python/22-2.py
import re

def intersect(a, b):
    return (max(a[0], b[0]), min(a[1], b[1]),
            max(a[2], b[2]), min(a[3], b[3]),
            max(a[4], b[4]), min(a[5], b[5]))

def overlaps(a, b):
    return not (a[1] < b[0] or a[0] > b[1] or
                a[3] < b[2] or a[2] > b[3] or
                a[5] < b[4] or a[4] > b[5])

def area(c):
    return (c[1] - c[0] + 1) * (c[3] - c[2] + 1) * (c[5] - c[4] + 1)

with open('input/22.txt') as f:
    data = [(line[1]=='n', [int(x) for x in re.findall('-?\d+', line)]) for line in f]

def solve(part1):
    cubes = []
    bounds = (-50,50,-50,50,-50,50)
    for OP, a in data:
        if part1 and not overlaps(bounds, a): continue
        new_cubes = []
        for b, w in cubes:
            if overlaps(a, b):
                new_cubes.append((intersect(a, b), -w))
        if OP:
            new_cubes.append((a, 1))
        cubes += new_cubes

    return sum(area(c)*w for c,w in cubes)

print("Part 1:", solve(True))
print("Part 2:", solve(False))

exit(0)
"""
Horrible, code golfed version:
"""
for L,(O,a)in enumerate([K:=list(open('input/22.txt'))][:0]+[(l[1]=='n',[int(x)for x in __import__('re').findall('-?\d+',l)])for l in K]+[C:=[]][:0]):(C:=C+[((max(a[0],b[0]),min(a[1],b[1]),max(a[2],b[2]),min(a[3],b[3]),max(a[4],b[4]),min(a[5],b[5])),-w) for b,w in C if (a[1]>=b[0]and a[0]<=b[1]and a[3]>=b[2]and a[2]<=b[3]and a[5]>=b[4]and a[4]<=b[5])]+[(a,1)]*O)and(L==len(K)-1)and(print(sum(((c[1]-c[0]+1)*(c[3]-c[2]+1)*(c[5]-c[4]+1))*w for c,w in C)))
